src/graph/immutable_graph.cc

```diff
@@ -197,6 +215,44 @@ std::pair<ImmutableGraph::CSR::Ptr, IdArray> ImmutableGraph::CSR::VertexSubgraph
   return std::pair<ImmutableGraph::CSR::Ptr, IdArray>(sub_csr, rst_eids);
 }
 
+std::pair<ImmutableGraph::CSR::Ptr, IdArray> ImmutableGraph::CSR::EdgeSubgraph(
+    IdArray eids, EdgeList::Ptr edge_list) const {
+  // Return sub_csr and vids array.
+  CHECK(IsValidIdArray(eids)) << "Invalid edge id array.";
+  const dgl_id_t* eid_data = static_cast<dgl_id_t*>(eids->data);
+  const int64_t len = eids->shape[0];
+  std::vector<dgl_id_t> nodes;
+  std::unordered_map<dgl_id_t, dgl_id_t> oldv2newv;
+
+  for (int64_t i = 0; i < len; i++) {
+    dgl_id_t src_id = edge_list->src_points[eid_data[i]];
+    dgl_id_t dst_id = edge_list->dst_points[eid_data[i]];
+    if (oldv2newv.insert(std::make_pair(src_id, oldv2newv.size())).second)
+      nodes.push_back(src_id);
+    if (oldv2newv.insert(std::make_pair(dst_id, oldv2newv.size())).second)
+      nodes.push_back(dst_id);
+  }
+
+  const size_t n = oldv2newv.size();
+
+  std::vector<Edge> edges;
+  for (int64_t i = 0; i < len; i++) {
+    dgl_id_t src_id = oldv2newv.find(edge_list->src_points[eid_data[i]])->second;
+    dgl_id_t dst_id = oldv2newv.find(edge_list->dst_points[eid_data[i]])->second;
+    edges.push_back(Edge{src_id, dst_id, static_cast<dgl_id_t>(i)});
+  }
```

Comment:
why do you not merge this loop with the one above. we should avoid hashtable lookups as much as possible.